### PR TITLE
Add QA-375 for OVS restart

### PR DESCRIPTION
**Pull request by alex7c4**

### CODE REVIEW

File: mos_tests/neutron/python_tests/base.py [hunk 1]
```diff
@@ -126,7 +126,7 @@ def execute():
 
     def check_ping_from_vm(self, vm, vm_keypair=None, ip_to_ping=None,
                            timeout=3 * 60, vm_login='cirros',
-                           vm_password='cubswin:)'):
+                           vm_password='cubswin:)', should_be_available=True):
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/base.py]: May be better make 3 methods:
1. run_ping_from_vm (which return exit code)
2. check_ping_from_vm (which call 1st and check exit code is 0)
3. check_no_ping_from_vm (which call 1st and check exit code is not 0)

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 2]
```diff
@@ -68,7 +69,10 @@ def disable_ovs_agents_on_controller(self):
             result = remote.execute(
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: `result = remote.check_call`

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 3]
```diff
@@ -78,7 +82,10 @@ def restart_ovs_agents_on_computes(self):
             with node.ssh() as remote:
                 result = remote.execute(
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: `result = remote.check_call`

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 4]
```diff
@@ -89,6 +96,10 @@ def enable_ovs_agents_on_controllers(self):
                 '. openrc && pcs resource enable '
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:94]: `result = remote.check_call`

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 5]
```diff
@@ -895,3 +906,126 @@ def test_ovs_new_flows_added_after_restart(self):
 
         after_value = self.get_current_cookie(compute)
         assert before_value != after_value
+
+
+@pytest.mark.check_env_("has_2_or_more_computes")
+@pytest.mark.check_env_("is_vlan")
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: `@pytest.mark.check_env_("has_2_or_more_computes", "is_vlan")`
single decorator

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 6]
```diff
@@ -895,3 +906,126 @@ def test_ovs_new_flows_added_after_restart(self):
...
+
+        zone = self.os_conn.nova.availability_zones.find(zoneName="nova")
+        hosts = zone.hosts.keys()[:2]
+
+        # Try to add sec groups if they were not added before
+        logger.info('Try to add sec groups if they were not added before')
+        try:
+            self.setup_rules_for_default_sec_group()
+        except Exception as e:
+            logger.warning('Warning during security rule update:'
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: exception is normal in this case?

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 7]
```diff
@@ -895,3 +906,126 @@ def test_ovs_new_flows_added_after_restart(self):
...
+                           '\n{0}'.format(e))
+
+        # Create two routers for two configs
+        logger.info('Create two routers')
+        router_05 = self.os_conn.create_router(name="test_router_05")
+        router_06 = self.os_conn.create_router(name="test_router_06")
+
+        # Create 2 separate networks and 2 vm instances
+        # and associate each element to their router (06net+06sub-> 06 router)
+        for i, hostname in enumerate(hosts, 5):
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: let's create routers inside loop

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 8]
```diff
@@ -895,3 +906,126 @@ def test_ovs_new_flows_added_after_restart(self):
...
+        1. Update default security group if needed
+        2. Create CONFIG 1:
+        Network: test_net_05
+        SubNetw: test_net_05__subnet, 192.168.5.0/24
+        Router:  test_router_05
+        3. Create CONFIG 2:
+        Network: test_net_06
+        SubNetw: test_net_06__subnet, 192.168.6.0/24
+        Router:  test_router_06
+        3. Launch 'test_vm_05' inside 'config 1'
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: step 3 after step 3

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 9]
```diff
@@ -895,3 +906,126 @@ def test_ovs_new_flows_added_after_restart(self):
...
+        Pings should NOT go between VMs.
+        5. Disable ovs-agents on all controllers, restart service
+        neutron-plugin-openvswitch-agent on all computes, and enable
+        them back. To do this, launch the script against master node.
+        6. Wait 30 seconds, send pings from 'test_vm_05' to 'test_vm_06'
+        and check that they are still NOT successful.
+
+        Duration 10m
+
+        https://mirantis.testrail.com/index.php?/cases/view/542666
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: this links are not public

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 10]
```diff
@@ -895,3 +906,126 @@ def test_ovs_new_flows_added_after_restart(self):
...
+        https://mirantis.jira.com/browse/QA-375
+        """
+        self.disable_ovs_agents_on_controller()
+        self.restart_ovs_agents_on_computes()
+        self.enable_ovs_agents_on_controllers()
+
+        # sleep is used to check that system will be stable for some time
+        # after restarting service
+        time.sleep(30)
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:1046]: why 30 seconds? Do you wait some event?

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 7 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: Why do we need a loop to create just two routers?

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 10 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:1046]: I'll add additional checks here, but anyway - put 30sec sleep was a recommendation from MOS team to have more stable env after agents restart

File: mos_tests/neutron/python_tests/base.py (hunk 1 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/base.py]: I think one universal ping method should be ok

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 7 above)
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: Because this change will make code shorter and clearer (without if inside loop)

Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: Ok, I thought you were talking about creation of separate loop

File: mos_tests/neutron/python_tests/base.py (hunk 1 above)
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/base.py]: I believe, that in code call 

```
self.check_no_ping_from_vm(*args, **kwargs)
```

looks clearer than

```
self.check_ping_from_vm(*args, should_be_available=False,**kwargs)
```

Comment (alex7c4) [inline mos_tests/neutron/python_tests/base.py]: Ok, I'll add separate NO ping method

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 4 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:94]: done

File: mos_tests/neutron/python_tests/base.py [hunk 11]
```diff
@@ -141,6 +141,26 @@ def check_ping_from_vm(self, vm, vm_keypair=None, ip_to_ping=None,
...
+        res = self.run_on_vm(vm, vm_keypair, cmd, timeout=timeout,
+                             vm_login=vm_login, vm_password=vm_password)
+        error_msg = (
+            'Instance HAS connection, but it should not.\n'
+            'EXIT CODE: "{exit_code}"\n'
+            'STDOUT: "{stdout}"\n'
+            'STDERR {stderr}').format(**res)
+        # Ping should not pass
+        logger.info('Expecting that ping from vm should fail')
+        assert 1 == res['exit_code'], error_msg
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/base.py]: What if exit_code will be 2?

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 12]
```diff
@@ -65,10 +66,12 @@ def disable_ovs_agents_on_controller(self):
...
 
         with controller.ssh() as remote:
-            result = remote.execute(
+            result = remote.check_call(
                 '. openrc && pcs resource disable '
                 'p_neutron-plugin-openvswitch-agent --wait')
-            assert result['exit_code'] == 0
+            cmd_exit_code = result['exit_code']
+            if cmd_exit_code != 0:
+                logger.error(result['stderr'])
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:74]: this line never executed

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 13]
```diff
@@ -76,19 +79,23 @@ def restart_ovs_agents_on_computes(self):
 
         for node in computes:
             with node.ssh() as remote:
-                result = remote.execute(
+                result = remote.check_call(
                     'service neutron-plugin-openvswitch-agent restart')
-                assert result['exit_code'] == 0
+                cmd_exit_code = result['exit_code']
+                if cmd_exit_code != 0:
+                    logger.error(result['stderr'])
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:86]: this line never executed

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 14]
```diff
@@ -76,19 +79,23 @@ def restart_ovs_agents_on_computes(self):
...
 
         with controller.ssh() as remote:
-            result = remote.execute(
+            result = remote.check_call(
                 '. openrc && pcs resource enable '
                 'p_neutron-plugin-openvswitch-agent --wait')
-            assert result['exit_code'] == 0
+            cmd_exit_code = result['exit_code']
+            if cmd_exit_code != 0:
+                logger.error(result['stderr'])
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:98]: this line never executed

File: mos_tests/neutron/python_tests/base.py [hunk 15]
```diff
@@ -141,6 +141,26 @@ def check_ping_from_vm(self, vm, vm_keypair=None, ip_to_ping=None,
         ).format(**res)
         assert 0 == res['exit_code'], error_msg
 
+    def check_no_ping_from_vm(self, vm, vm_keypair=None, ip_to_ping=None,
+                              timeout=3 * 60, vm_login='cirros',
+                              vm_password='cubswin:)'):
+        if ip_to_ping is None:
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/base.py:167]: Let's extract common logic (L130-L136, and L147-L153) to new method and call them from both `check_no_ping_from_vm` and `check_ping_from_vm`

File: mos_tests/neutron/python_tests/test_ovs_restart.py [hunk 16]
```diff
@@ -895,3 +902,148 @@ def test_ovs_new_flows_added_after_restart(self):
...
+
+        # sleep is used to check that system will be stable for some time
+        # after restarting service
+        time.sleep(30)
+
+        self.check_no_ping_from_vm(self.server1, self.instance_keypair,
+                                   self.server2_ip, timeout=None)
+
+        # check all agents are alive
+        assert all([agt['alive'] for agt in
```
Comment (gdyuldin) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: That step is not necessary for steps (from docstring). May be you need to add 9th step?

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 2 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py]: done

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 12 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:74]: done

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 13 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:86]: done

File: mos_tests/neutron/python_tests/test_ovs_restart.py (hunk 14 above)
Comment (alex7c4) [inline mos_tests/neutron/python_tests/test_ovs_restart.py:98]: done

**Pull request merged**